controllers/custom/builder.go

```diff
@@ -0,0 +1,201 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package custom
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+type Builder struct {
+	// options has the configurable parameters for the custom controller
+	options Options
+	// converter which will be used to convert a k8s object into desired
+	// format before it's stored in the data store
+	converter Converter
+	// clientSet is the kubernetes client set
+	clientSet *kubernetes.Clientset
+	// dataStore with the converted k8s object, objects being watched by the
+	// controller must be queried using this datastore
+	dataStore cache.Indexer
+	// mgr is the controller runtime manager
+	mgr               manager.Manager
+	dataStoreSyncFlag *bool
+
+	log logr.Logger
+}
+
+func (b *Builder) Named(name string) *Builder {
+	b.options.Name = name
+	return b
+}
+
+func (b *Builder) UsingConverter(converter Converter) *Builder {
+	b.converter = converter
+	return b
+}
+
+func (b *Builder) WithClientSet(clientSet *kubernetes.Clientset) *Builder {
+	b.clientSet = clientSet
+	return b
+}
+
+func (b *Builder) UsingDataStore(dataStore cache.Indexer) *Builder {
+	b.dataStore = dataStore
+	return b
+}
+
+func (b *Builder) WithLogger(logger logr.Logger) *Builder {
+	b.log = logger
+	return b
+}
+
+func (b *Builder) Options(options Options) *Builder {
+	b.options = options
+	return b
+}
+
+func (b *Builder) DataStoreSyncFlag(flag *bool) *Builder {
+	b.dataStoreSyncFlag = flag
+	return b
+}
+
+func NewControllerManagedBy(mgr manager.Manager) *Builder {
+	return &Builder{mgr: mgr}
+}
+
+// Complete adds the controller to manager's Runnable. The Controller
+// runnable will start when the manager starts
+func (b *Builder) Complete(reconciler Reconciler) error {
```

Comment:
`Complete` sems very generic a name.. How about something specific like `AddController` or anything which is specific.

Reply:
I tried to keep it inline with upstream controller's interface -  https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.9/pkg/builder/controller.go#L167-L171

I am fine with renaming it to AddController too.

Reply:
Consistency with upstream is better. So, +1 in keeping it so.
I didn't realize that both are enabling a builder pattern. I looked at other methods and then realized it.